pkg/controller/sls/project_controller.go

```diff
@@ -0,0 +1,130 @@
+
+	slsv1alpha1 "github.com/crossplane/provider-alibaba/apis/sls/v1alpha1"
+	"github.com/crossplane/provider-alibaba/apis/v1alpha1"
+	slsclient "github.com/crossplane/provider-alibaba/pkg/clients/sls"
+)
+
+const (
+	errNotSLSProject            = "managed resource is not an SLS project custom resource"
+	errNoProvider               = "no provider config or provider specified"
+	errGetProviderConfig        = "cannot get provider config"
+	errTrackUsage               = "cannot track provider config usage"
+	errNoConnectionSecret       = "no connection secret specified"
+	errGetConnectionSecret      = "cannot get connection secret"
+	errFmtUnsupportedCredSource = "credentials source %q is not currently supported"
+)
+
+// SetupSLSProject adds a controller that reconciles SLSProjects.
+func SetupSLSProject(mgr ctrl.Manager, l logging.Logger) error {
+	options := []managed.ReconcilerOption{
+		managed.WithExternalConnecter(&connector{
+			client:      mgr.GetClient(),
+			usage:       resource.NewProviderConfigUsageTracker(mgr.GetClient(), &v1alpha1.ProviderConfigUsage{}),
+			NewClientFn: slsclient.NewClient,
+		})}
+
+	return BaseSetupSLSProject(mgr, l, options...)
+}
+
+type connector struct {
+	client      client.Client
+	usage       resource.Tracker
+	NewClientFn func(accessKeyID, accessKeySecret, region string) *slsclient.LogClient
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) { //nolint:gocyclo
+	cr, ok := mg.(*slsv1alpha1.SLSProject)
+	if !ok {
+		return nil, errors.New(errNotSLSProject)
+	}
+
+	// provider has more than one kind of managed resource.
+	var (
+		sel    *xpv1.SecretKeySelector
+		region string
+	)
+
+	switch {
+	case cr.GetProviderConfigReference() != nil:
+		if err := c.usage.Track(ctx, mg); err != nil {
+			return nil, errors.Wrap(err, errTrackUsage)
+		}
+
+		pc := &v1alpha1.ProviderConfig{}
+		if err := c.client.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderConfigReference.Name}, pc); err != nil {
+			return nil, errors.Wrap(err, errGetProviderConfig)
+		}
+		if s := pc.Spec.Credentials.Source; s != xpv1.CredentialsSourceSecret {
+			return nil, errors.Errorf(errFmtUnsupportedCredSource, s)
+		}
+		sel = pc.Spec.Credentials.SecretRef
+		region = pc.Spec.Region
+	default:
+		return nil, errors.New(errNoProvider)
+	}
+
+	if sel == nil {
+		return nil, errors.New(errNoConnectionSecret)
+	}
+
+	s := &corev1.Secret{}
+	nn := types.NamespacedName{Namespace: sel.Namespace, Name: sel.Name}
+	if err := c.client.Get(ctx, nn, s); err != nil {
+		return nil, errors.Wrap(err, errGetConnectionSecret)
+	}
+
+	slsClient := c.NewClientFn(string(s.Data["accessKeyId"]), string(s.Data["accessKeySecret"]), region)
+	return &external{client: slsClient}, nil
+}
+
+type external struct {
+	client slsclient.LogClientInterface
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	return BaseObserve(mg, e.client)
```

Comment:
What is gained by breaking all this logic out into these separate `Base` functions, compared to just writing them inline?

Reply:
Please kindly refer to issue #63, in which I tried to explain the reason in details and wrote a few examples.